pkg/downwardmetrics/virtio-serial/server.go

```diff
@@ -0,0 +1,276 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2023 Red Hat, Inc.
+ *
+ */
+
+package virtio_serial
+
+import (
+	"bufio"
+	"encoding/xml"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/textproto"
+	"net/url"
+	"strings"
+	"syscall"
+	"time"
+
+	"kubevirt.io/client-go/log"
+
+	"kubevirt.io/kubevirt/pkg/downwardmetrics"
+	"kubevirt.io/kubevirt/pkg/downwardmetrics/vhostmd/api"
+	"kubevirt.io/kubevirt/pkg/util"
+	"kubevirt.io/kubevirt/pkg/virt-handler/isolation"
+
+	diskutils "kubevirt.io/kubevirt/pkg/ephemeral-disk-utils"
+	metricsScraper "kubevirt.io/kubevirt/pkg/monitoring/domainstats/downwardmetrics"
+)
+
+const connectAttempts = 6
+const standardLauncherSocketPath = "/run/kubevirt/sockets/launcher-sock"
+
+var (
+	invalidRequest = []byte("INVALID REQUEST\n\n")
+	emptyMetrics   = []byte("<metrics><!-- host metrics not available --><!-- VM metrics not available --></metrics>")
+)
+
+func RunDownwardMetricsVirtioServer(nodeName string, stop chan struct{}) error {
+	report, err := newMetricsReporter(nodeName)
+	if err != nil {
+		return err
+	}
+
+	server := downwardMetricsServer{
+		connectAttempts:    connectAttempts,
+		virtioSerialSocket: downwardmetrics.DownwardMetricsChannelSocket,
+		report:             report,
+	}
+	return server.start(stop)
+}
+
+type metricsReporter func() (*api.Metrics, error)
+
+func newMetricsReporter(nodeName string) (metricsReporter, error) {
+	exists, err := diskutils.FileExists(standardLauncherSocketPath)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, errors.New("virt-launcher socket not found")
+	}
+
+	isolationDetector := isolation.NewSocketBasedIsolationDetector(util.VirtShareDir)
+	scraper := metricsScraper.NewScraper(nodeName, isolationDetector)
+
+	return func() (*api.Metrics, error) {
+		return scraper.Report(standardLauncherSocketPath)
+	}, nil
+}
+
+type downwardMetricsServer struct {
+	connectAttempts    uint
+	virtioSerialSocket string
+	report             metricsReporter
+}
+
+func (s *downwardMetricsServer) start(stop chan struct{}) error {
+	conn, err := connect(s.virtioSerialSocket, s.connectAttempts)
+	if err != nil {
+		return err
+	}
+
+	go s.serve(conn, stop)
+	return nil
+}
+
+func (s *downwardMetricsServer) serve(conn net.Conn, stop chan struct{}) {
+	defer func(conn net.Conn) {
+		err := conn.Close()
+		if err != nil {
+			log.Log.Reason(err).Warning("virtio-serial failed to close the connection")
+		}
+	}(conn)
+
+	type reqResult struct {
+		request string
+		err     error
+	}
+	newRequest := make(chan reqResult)
+	defer close(newRequest)
+	reader := textproto.NewReader(bufio.NewReader(conn))
+
+	for {
```

Comment:
IMO it is cleaner to have separate go routines for 1.) running the processing of incoming requests and 2.) for the collection of metrics + sending the responses out again.

Reply:
because of how virtio-serial works (it's not connection oriented), we can serve one request at a time. So I don't think it would be clearer to spawn a new goroutine and wait until it finished before serve the next request.

Reply:
sadly we can't use vsock (it will make this code more simple) since SAP only supports virtio-serial

Reply:
It might be still more efficient to place the receiving routine outside the processing for-loop and adding a for-loop into it? This way it won't be respawned every iteration of the processing loop?

Reply:
The problem in that case is this `request, err := reader.ReadLine()` is a blocking call, so if we do that we don't have a mean to stop the server if the channel `stop` is closed. I could do that using two for-loops, and an extra channel to not call `waitForRequest()` until the previous request has been served. I don't think the complexity is worth it to avoid having to spawn a new goroutine, specially how this will be used by SAP

Reply:
Don't you have this problem anyway? How do you stop the routine waiting for a request if it is blocked? Will it exit when the `conn` is closed?

Reply:
Yes, exactly. The "main" gorutine (i.e., `go serve()`) closes `conn` when the `stop` channel is closed, so `reader.ReadLine()` (inside`waitForRequest()`) will unblock returning `net.ErrClosed`

Reply:
Just to clarify, wouldn't the go-routine do the same if it was running its own for-loop?